lib/libtsd/tsd_task.3
.\"-
.\" Copyright (c) 2015 The University of Oslo
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. The name of the author may not be used to endorse or promote
.\"    products derived from this software without specific prior written
.\"    permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd February 5, 2015
.Dt TSD_TASK 3
.Os
.Sh NAME
.Nm tsd_task_create ,
.Nm tsd_task_destroy ,
.Nm tsd_task_setuser ,
.Nm tsd_task_setcred ,
.Nm tsd_task_start ,
.Nm tsd_task_stop ,
.Nm tsd_task_signal ,
.Nm tsd_task_poll ,
.Nm tsd_task_reset
.Nd task management
.Sh LIBRARY
.Lb libtsd
.Sh SYNOPSIS
.In sys/types.h
.In tsd/task.h
.Ft typedef "void \*(lp*tsd_task_func\*(rp \*(lpvoid *\*(rp" ;
.Pp
.Ft struct tsd_task *
.Fn tsd_task_create "const char *name" "tsd_task_func *func" "void *data"
.Ft void
.Fn tsd_task_destroy "struct tsd_task *task"
.Ft int
.Fn tsd_task_setuser "struct tsd_task *task" "const char *user"
.Ft int
.Fn tsd_task_setcred "struct tsd_task *task" "uid_t uid" "gid_t *gids" "int ngids"
.Ft int
.Fn tsd_task_start "struct tsd_task *task"
.Ft int
.Fn tsd_task_stop "struct tsd_task *task"
.Ft int
.Fn tsd_task_signal "struct tsd_task *task" "int sig"
.Ft int
.Fn tsd_task_poll "struct tsd_task *task"
.Ft int
.Fn tsd_task_reset "struct tsd_task *task"
.Sh DESCRIPTION
The
.Nm tsd_task
API is used to create and manage tasks to be run in subprocesses.
The companion
.Xr tsd_task_set 3
and
.Xr tsd_task_queue 3
APIs can be used in addition to the
.Nm tsd_task
API to manage task sets and queues.
.Pp
The
.Fn tsd_task_create
function creates a new task.
The name can be an arbitrary string of up to 64 characters including
the terminating NUL, and is mostly decorative, unless the task is
assigned to a task set (see
.Xr tsd_task_set 3 Ns ),
in which case it must be unique within a task set.
The
.Va func
argument points to a function which will be called within the child
process when the task is started.
This function will normally execute an external program using
.Xr execve 2
or similar.
The final argument,
.Va data ,
points to caller-defined data and is passed unmodified to the child
function when the task is started.
.Pp
The
.Fn tsd_task_destroy
function destroys the specified task.
If the task is running, it is first stopped.
It is then removed from any sets and / or queues to which it may
belong.
Finally, all resources allocated by the
.Nm tsd_task
API on the task's behalf are released.
.Pp
The
.Fn tsd_task_setcred
and
.Fn tsd_task_setuser
functions are used to specify credentials with which the task should
run.
These credentials are set before the child function is invoked.
The
.Fn tsd_task_setcred
function is used to pass a UID and a set of GIDs directly, while the
.Fn tsd_task_setuser
function is used to pass the name of a user whose credentials should
be used.
.Pp
The
.Fn tsd_task_start
function starts the given task.
It sets up any requested pipes, forks a child process, sets the
requested credentials and finally invokes the child function.
.Pp
The
.Fn tsd_task_stop
function stops the given task.
It first checks whether the child process has already terminated.
If that is not the case, it will send
.Dv SIGCONT ,
.Dv SIGTERM
and
.Dv SIGKILL ,
in that order, with a brief pause after each signal to give the child
process time to react.
.\" XXX should we mark the task as TASK_KILLED instead of TASK_DEAD,
.\" XXX so we can still reap it if it wakes up and dies later, perhaps
.\" XXX due to having been swapped out on a heavily loaded system?
.Pp
The
.Fn tsd_task_signal
function sends the specified signal to the child process.
.Pp
The
.Fn tsd_task_poll
function polls a task to see if it is still running.
If not, it reaps the child process and sets the task state to reflect
the outcome (see the
.Sx TASK STATES
section below).
.Pp
The
.Fn tsd_task_reset
function resets a task to the
.Dv TASK_IDLE
(runnable) state after it has stopped so it can be started again in
the future.
.Sh TASK STATES
The
.Va state
member of
.Vt struc tsd_task
indicates the current state of the task.
The following states are defined:
.Bl -tag -width "TASK_FINISHED"
.\" .It Dv TASK_INVALID
.\" The task is in an inconsistent state.
.\" Defined but not used.
.It Dv TASK_IDLE
The task is idle and can be started.
.\" .It Dv TASK_QUEUED
.\" The task is queued to be started as soon as there is an available
.\" slot.
.\" See
.\" .Xr tsd_tqueue 3 .
.\" Defined but not used.
.It Dv TASK_STARTING
The task is in the process of starting.
This state is only used internally and should not be visible to the
caller.
.It Dv TASK_RUNNING
The task is currently running.
.It Dv TASK_STOPPING
The task is in the process of stopping.
This state is only used internally and should not be visible to the
caller.
.It Dv TASK_STOPPED
The task has stopped.
It terminated normally and returned an exit code of 0.
.It Dv TASK_DEAD
An error occurred within
.Fn tsd_task_start ,
.Fn tsd_task_poll
or
.Fn tsd_task_stop ,
or the task was terminated by a signal.
In the latter case, the
.Va status
member of
.Vt struct tsd_task
contain the exit status of the child process as returned by the
.Xr wait 2
family of system calls.
.\" XXX we should have separate states for these
.It Dv TASK_FINISHED
This state is not used internally, but can be used by the caller to
mark a task as having completed successfully.
.\" XXX that's dumb
.It Dv TASK_FAILED
The task terminated normally and returned a non-zero exit code.
The
.Va status
member of
.Vt struct tsd_task
contain the exit status of the child process as returned by the
.Xr wait 2
family of system calls.
.El
.Sh RETURN VALUES
TBW
.Sh IMPLEMENTATION NOTES
The
.Fn tsd_task_stop
function cannot be relied upon to terminate a child cleanly and should
be used on a best-effort basis as a last resort.
The preferred method is to use a pre-established communication channel
to somehow instruct the child to finish its work, and call
.Fn tsd_task_poll
regularly until it reports that the task has stopped.
.Sh BUGS
The use of pipes to communicate with the child process is mentioned
but not described.
.Sh SEE ALSO
.Xr execve 2 ,
.Xr fork 2 ,
.Xr kill 2 ,
.Xr pipe 2 ,
.Xr tsd_task_queue 3 ,
.Xr tsd_task_set 3
.Sh AUTHORS
The
.Fn tsd_task
API and this manual page were written for the University of Oslo by
.An Dag-Erling Sm\(/orgrav Aq Mt d.e.smorgrav@usit.uio.no .


lib/libtsd/tsd_task.c
/*-
 * Copyright (c) 2014-2015 The University of Oslo
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote
 *    products derived from this software without specific prior written
 *    permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <sys/types.h>
#include <sys/wait.h>

#include <errno.h>
#include <fcntl.h>
#include <grp.h>
#include <limits.h>
#include <pwd.h>
#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#if HAVE_BSD_UNISTD_H
#include <bsd/unistd.h>
#endif

#include <tsd/hash.h>
#include <tsd/log.h>
#include <tsd/strutil.h>
#include <tsd/task.h>

/*
 * Clear task credentials.
 */
static void
tsd_task_clearcred(struct tsd_task *t)
{

	memset(t->user, 0, sizeof t->user);
	t->uid = (uid_t)-1;
	memset(t->gids, 0, sizeof t->gids);
	t->gids[0] = (gid_t)-1;
	t->ngids = 0;
}

/*
 * Create a new task.
 */
struct tsd_task *
tsd_task_create(const char *name, tsd_task_func *func, void *ud)
{
	struct tsd_task *t;

	if ((t = calloc(1, sizeof *t)) == NULL)
		return (NULL);
	if (strlcpy(t->name, name, sizeof t->name) >= sizeof t->name) {
		memset(t, 0, sizeof *t);
		free(t);
		errno = ENAMETOOLONG;
		return (NULL);
	}
	t->h = tsd_strhash(t->name);
	t->state = TASK_IDLE;
	tsd_task_clearcred(t);
	t->func = func;
	t->pid = -1;
	t->pin = t->pout = t->perr = -1;
	t->ud = ud;
	VERBOSE("%s", name);
	return (t);
}

/*
 * Destroy a task.
 */
void
tsd_task_destroy(struct tsd_task *t)
{
	if (t == NULL)
		return;
	VERBOSE("%s", t->name);
	if (t->state == TASK_RUNNING)
		tsd_task_stop(t);
	if (t->queue != NULL)
		tsd_tqueue_remove(t->queue, t);
	if (t->set != NULL)
		tsd_tset_remove(t->set, t);
	/* ASSERT(t->state != TASK_STOPPING); */
	memset(t, 0, sizeof *t);
	free(t);
}

/*
 * Internal: perform cleanup after a task stops or fails.
 */
static void
tsd_task_close(struct tsd_task *t, enum tsd_task_state nextstate)
{

	t->pid = -1;
	if (t->flags & TASK_STDIN)
		close(t->pin);
	if (t->flags & TASK_STDOUT)
		close(t->pout);
	if (t->flags & TASK_STDERR)
		close(t->perr);
	t->pin = t->pout = t->perr = -1;
	if (t->state == TASK_RUNNING || t->state == TASK_STOPPING) {
		/* currently counted as running */
		if (t->set != NULL)
			t->set->nrunning--;
		if (t->queue != NULL)
			tsd_tqueue_remove(t->queue, t);
	}
	t->state = nextstate;
}

/*
 * Set the task credentials to those of the given user.
 */
int
tsd_task_setuser(struct tsd_task *t, const char *user)
{
	struct passwd *pwd;

	if (t->state != TASK_IDLE) {
		errno = EBUSY;
		return (-1);
	}
	errno = 0;
	if ((pwd = getpwnam(user)) == NULL) {
		if (errno == 0)
			errno = ENOENT;
		goto fail;
	}
	if (strlcpy(t->user, pwd->pw_name, sizeof t->user) >= sizeof t->user) {
		errno = ENAMETOOLONG;
		goto fail;
	}
	t->uid = pwd->pw_uid;
	t->ngids = sizeof t->gids / sizeof t->gids[0];
	if (getgrouplist(pwd->pw_name, pwd->pw_gid, t->gids, &t->ngids) < 0) {
		/* XXX does getgrouplist() set errno? */
		goto fail;
	}
	return (0);
fail:
	tsd_task_clearcred(t);
	return (-1);
}

/*
 * Set the task credentials to the given UID and GID.
 */
int
tsd_task_setcred(struct tsd_task *t, uid_t uid, gid_t *gids, int ngids)
{

	if (t->state != TASK_IDLE) {
		errno = EBUSY;
		return (-1);
	}
	tsd_task_clearcred(t);
	if (ngids < 1 || ngids > (int)(sizeof t->gids / sizeof t->gids[0]))
		return (-1);
	snprintf(t->user, sizeof t->user, "(%lu:%lu)",
	    (unsigned long)uid, (unsigned long)gids[0]);
	t->uid = uid;
	memcpy(t->gids, gids, ngids * sizeof *gids);
	t->ngids = ngids;
	return (0);
}

/*
 * Fork a child process and start a task inside it.
 *
 * XXX consider adding support for chroot
 */
int
tsd_task_start(struct tsd_task *t)
{
	int pin[2] = { -1, -1 };
	int pout[2] = { -1, -1 };
	int perr[2] = { -1, -1 };
	int ret, serrno;
#if !HAVE_CLOSEFROM
	int fd, maxfd;
#endif

	VERBOSE("%s", t->name);

	if (t->state == TASK_RUNNING)
		return (0);
	if (t->state != TASK_IDLE)
		return (-1);
	t->state = TASK_STARTING;

	/* prepare file descriptors */
	if (t->flags & TASK_STDIN_NULL) {
		if ((pin[0] = open("/dev/null", O_RDONLY)) < 0)
			goto fail;
	} else if (t->flags & TASK_STDIN_PIPE) {
		if (pipe(pin) != 0 || fcntl(pin[1], F_SETFL, O_NONBLOCK) != 0)
			goto fail;
		t->pin = pin[1];
	}
	if (t->flags & TASK_STDOUT_NULL) {
		if ((pout[1] = open("/dev/null", O_WRONLY)) < 0)
			goto fail;
	} else if (t->flags & TASK_STDOUT_PIPE) {
		if (pipe(pout) != 0 || fcntl(pout[0], F_SETFL, O_NONBLOCK) != 0)
			goto fail;
		t->pout = pout[0];
	}
	if (t->flags & TASK_STDERR_NULL) {
		if ((perr[1] = open("/dev/null", O_WRONLY)) < 0)
			goto fail;
	} else if (t->flags & TASK_STDERR_PIPE) {
		if (pipe(perr) != 0 || fcntl(perr[0], F_SETFL, O_NONBLOCK) != 0)
			goto fail;
		t->perr = perr[0];
	}

	/* fork the child */
	fflush(NULL);
	if ((t->pid = fork()) < 0)
		goto fail;

	/* child */
	if (t->pid == 0) {
		/* set up stdin/out/err and close everything else */
#if HAVE_FPURGE
		fpurge(stdin);
#endif
		if ((t->flags & TASK_STDIN && dup2(pin[0], 0) != 0) ||
		    (t->flags & TASK_STDOUT && dup2(pout[1], 1) != 1) ||
		    (t->flags & TASK_STDERR && dup2(perr[1], 2) != 2)) {
			ERROR("failed to set up standard file descriptors");
			_exit(1);
		}
#if HAVE_CLOSEFROM
		closefrom(3);
#else
		maxfd = getdtablesize();
		for (fd = 3; fd < maxfd; ++fd)
			close(fd);
#endif

		/* set process title if possible */
#if HAVE_SETPROCTITLE
		setproctitle("%s", t->name);
#endif

		/* drop privileges */
		if (geteuid() == 0 && t->gids[0] > 0 && t->uid != (uid_t)-1) {
			if ((ret = setgid(t->gids[0])) != 0)
				ERROR("failed to set process group");
#if HAVE_SETGROUPS
			else if ((ret = setgroups(t->ngids, t->gids)) != 0)
				ERROR("failed to set additional process groups");
#endif
			else if ((ret = setuid(t->uid)) != 0)
				ERROR("failed to set process user");
			if (ret != 0)
				_exit(1);
		}
		if (getgid() != getegid())
			(void)setgid(getgid());
		if (getuid() != geteuid())
			(void)setuid(getuid());
		(*t->func)(t->ud);
		_exit(1);
	}

	/* parent */
	if (t->flags & TASK_STDIN)
		close(pin[0]);
	if (t->flags & TASK_STDOUT)
		close(pout[1]);
	if (t->flags & TASK_STDERR)
		close(perr[1]);
	t->state = TASK_RUNNING;
	if (t->set != NULL)
		t->set->nrunning++;
	if (t->queue != NULL)
		t->queue->nrunning++;
	return (0);
fail:
	serrno = errno;
	/* close the child side, tsd_task_close() will close the other */
	if (t->flags & TASK_STDIN)
		close(pin[1]);
	if (t->flags & TASK_STDOUT)
		close(pout[0]);
	if (t->flags & TASK_STDERR)
		close(perr[0]);
	tsd_task_close(t, TASK_DEAD);
	errno = serrno;
	return (-1);
}

/*
 * Stop a task.  We make three attempts: first we check to see if the
 * child is already dead.  Then we wait 10 ms and check again; if it still
 * isn't dead, we send a SIGTERM, wait 10 ms, and check again.  If it
 * still isn't dead after the SIGTERM, we send a SIGKILL, wait 10 ms, and
 * try one last time.
 */
int
tsd_task_stop(struct tsd_task *t)
{
	static const int sig[] = { SIGCONT, SIGTERM, SIGKILL, 0, -1 };
	int i, serrno;

	VERBOSE("%s", t->name);

	/* check current state */
	if (t->state != TASK_RUNNING)
		return (-1);
	t->state = TASK_STOPPING;

	/* reap the child */
	for (i = 0; sig[i] >= 0; ++i) {
		tsd_task_poll(t);
		if (t->state != TASK_STOPPING)
			break;
		/* not dead yet; kill, wait 100 ms and retry */
		if (kill(t->pid, sig[i]) != 0) {
			serrno = errno;
			WARNING("unable to signal child %d", (int)t->pid);
			tsd_task_close(t, TASK_DEAD);
			errno = serrno;
			return (-1);
		}
		usleep(100000);
	}

	/* either done or gave up */
	if (sig[i] < 0) {
		WARNING("gave up waiting for child %d", (int)t->pid);
		tsd_task_close(t, TASK_DEAD);
		/* XXX set errno? */
		return (-1);
	}

	/* in summary... */
	if (t->state != TASK_STOPPED)
		return (-1);
	return (0);
}

/*
 * Send a signal to a task.
 */
int
tsd_task_signal(const struct tsd_task *t, int sig)
{

	VERBOSE("%d", sig);

	if (t->state != TASK_RUNNING) {
		errno = ESRCH;
		return (-1);
	}
	return (kill(t->pid, sig));
}

/*
 * Reset a task so it can be started again.
 * Note: it is the caller's responsibility to add it to a queue.
 */
int
tsd_task_reset(struct tsd_task *t)
{

	VERBOSE("%s", t->name);

	if (t->state == TASK_IDLE)
		return (0);
	if (t->state == TASK_RUNNING)
		tsd_task_stop(t);
	t->status = 0;
	t->state = TASK_IDLE;
	return (0);
}

/*
 * Poll a task to see if it's still running.
 */
int
tsd_task_poll(struct tsd_task *t)
{
	enum tsd_task_state nextstate;
	int ret, serrno;

	VERBOSE("%s", t->name);

	if (t->state != TASK_RUNNING && t->state != TASK_STOPPING)
		return (-1);
	nextstate = t->state;
	ret = waitpid(t->pid, &t->status, WNOHANG);
	if (ret < 0) {
		serrno = errno;
		/* already reaped, or something is wrong */
		WARNING("waitpid(%lu): %s",
		    (unsigned long)t->pid, strerror(errno));
		nextstate = TASK_DEAD;
		errno = serrno;
		/* fall through */
	} else if (ret == 0) {
		/* still running */
		return (0);
	} else if (ret == t->pid) {
		if (WIFEXITED(t->status) && WEXITSTATUS(t->status) == 0) {
			VERBOSE("%s [%lu] succeeded", t->name,
			    (unsigned long)t->pid);
			nextstate = TASK_STOPPED;
		} else if (WIFEXITED(t->status)) {
			NOTICE("%s [%lu] failed with exit code %d", t->name,
			    (unsigned long)t->pid, WEXITSTATUS(t->status));
			nextstate = TASK_FAILED;
		} else if (WIFSIGNALED(t->status)) {
			WARNING("%s [%lu] caught signal %d", t->name,
			    (unsigned long)t->pid, WTERMSIG(t->status));
			nextstate = TASK_DEAD;
		} else {
			ERROR("%s [%lu] terminated with unknown status %d",
			    t->name, (unsigned long)t->pid, t->status);
			nextstate = TASK_DEAD;
		}
		/* fall through */
	} else {
		/* wtf? */
		ERROR("waitpid(%lu) returned %d",
		    (unsigned long)t->pid, ret);
		errno = EAGAIN;
		return (-1);
	}
	/* fell through from above: task has stopped */
	serrno = errno;
	tsd_task_close(t, nextstate);
	errno = serrno;
	if (t->state != TASK_STOPPED)
		return (-1);
	return (0);
}
